src/main/java/org/jenkins/tools/test/util/WarExtractor.java

```diff
@@ -0,0 +1,160 @@
+import java.util.jar.JarInputStream;
+import java.util.jar.Manifest;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import org.apache.maven.model.Model;
+import org.jenkins.tools.test.exception.MetadataExtractionException;
+import org.jenkins.tools.test.model.plugin_metadata.Plugin;
+import org.jenkins.tools.test.model.plugin_metadata.PluginMetadataExtractor;
+
+public class WarExtractor {
+
+    private static final Logger LOGGER = Logger.getLogger(WarExtractor.class.getName());
+
+    private static final String PREFIX = "WEB-INF/plugins/";
+
+    private static final String SUFFIX = ".hpi";
+
+    @NonNull
+    private final File warFile;
+
+    @NonNull
+    private final List<PluginMetadataExtractor> extractors;
+
+    @CheckForNull
+    private final Set<String> includedPlugins;
+
+    @CheckForNull
+    private final Set<String> excludedPlugins;
+
+    public WarExtractor(
+            File warFile, ServiceHelper serviceHelper, Set<String> includedPlugins, Set<String> excludedPlugins) {
+        this.warFile = warFile;
+        this.extractors = serviceHelper.loadServices(PluginMetadataExtractor.class);
+        this.includedPlugins = includedPlugins;
+        this.excludedPlugins = excludedPlugins;
+    }
+
+    /**
+     * Extract the Jenkins core version from the given WAR.
+     *
+     * @return The Jenkins core version..
+     */
+    public String extractCoreVersion() throws MetadataExtractionException {
+        try (JarFile jf = new JarFile(warFile)) {
+            Manifest manifest = jf.getManifest();
+            String value = manifest.getMainAttributes().getValue("Jenkins-Version");
+            if (value == null) {
+                throw new MetadataExtractionException("Jenkins WAR is missing required Manifest entry");
+            }
+            return value;
+        } catch (IOException e) {
+            throw new UncheckedIOException("Failed to extract Jenkins core version from " + warFile.toString(), e);
+        }
+    }
+
+    /**
+     * Extract the list of plugins to be tested from the given WAR.
+     *
+     * @return An unmodifiable list of plugins to be tested, sorted by plugin ID.
+     */
+    public List<Plugin> extractPlugins() throws MetadataExtractionException {
+        List<Plugin> plugins = new ArrayList<>();
+        try (JarFile jf = new JarFile(warFile)) {
+            Enumeration<JarEntry> entries = jf.entries();
+            while (entries.hasMoreElements()) {
+                JarEntry entry = entries.nextElement();
+                if (isInteresting(entry)) {
+                    plugins.add(getPlugin(jf, entry));
+                }
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException("I/O error occurred whilst extracting plugin metadata from WAR", e);
+        }
+        if (plugins.isEmpty()) {
+            throw new MetadataExtractionException("Found no plugins in " + warFile);
+        }
```

Comment:
this breaks the ability to run a PCT against a plugin that is in a local clone only.  this was previously working,